src/perf/best-practices.md

```diff
@@ -0,0 +1,411 @@
+app's UI can have a big impact on how efficiently it runs.
+
+## Control build() cost
+
+Here are some things to keep in mind when designing your UI:
+
+* Avoid repetitive and costly work in `build()` methods
+  since `build()` can be invoked frequently when
+  ancestor widgets rebuild.
+* Avoid overly large single widgets with a large `build()` function.
+  Split them into different widgets based on encapsulation
+  but also on how they change:
+  * When `setState()` is called on a `State` object,
+    all descendent widgets rebuild. Therefore,
+    localize the `setState()` call to the part of
+    the subtree whose UI actually needs to change.
+    Avoid calling `setState()` high up in the tree
+    if the change is contained to a small part of the tree.
+  * The traversal to rebuild all descendents stops when the
+    same instance of the child widget as the previous
+    frame is re-encountered. This technique is heavily
+    used inside the framework for optimizing
+    animations where the animation doesn't affect the child subtree.
+    See the [`TransitionBuilder`][] pattern and
+    the [source code for `SlideTransition`][],
+    which uses this principle to avoid rebuilding its
+    descendents when animating.
+
+Use `const` constructors on widgets as much as possible,
+since they allow Flutter to short-circuit most of the rebuild work.
+To create reusable pieces of UIs,
+prefer using a [`StatelessWidget`][] rather than a function.
+
+For more information, check out:
+
+* [Performance considerations][],
+  part of the [`StatefulWidget`][] API doc
+* [Widgets vs helper methods][],
+  a video from the official Flutter YouTube 
+  channel that explains why widgets
+  (especially widgets with `const` constructors)
+  are more performant than functions.
+
+[Performance considerations]: {{site.api}}/flutter/widgets/StatefulWidget-class.html#performance-considerations
+[source code for `SlideTransition`]: {{site.repo.flutter}}/blob/master/packages/flutter/lib/src/widgets/transitions.dart#L168
+[`StatefulWidget`]: {{site.api}}/flutter/widgets/StatefulWidget-class.html
+[`StatelessWidget`]: {{site.api}}/flutter/widgets/StatelessWidget-class.html
+[`TransitionBuilder`]: {{site.api}}/flutter/widgets/TransitionBuilder.html
+[Widgets vs helper methods]: {{site.youtube-site}}/watch?v=IOyq-eTRhvo
+
+---
+
+## Use saveLayer() thoughtfully
+
+Some Flutter code uses `saveLayer()`, an expensive operation,
+to implement various visual effects in the UI.
+Even if your code doesn't explicitly call `saveLayer()`,
+other widgets or packages that you use might call it behind the scenes.
+Perhaps your app is calling `saveLayer()` more than necessary.
+
+{{site.alert.secondary}}
+  <h4>Why is saveLayer expensive?</h4><a name="why-is-savelayer-expensive"></a>
+  Calling `saveLayer()` allocates an offscreen buffer.
+  Drawing content into the offscreen buffer might trigger
+  render target switches that are particularly slow in
+  older GPUs.
```

Comment:
I would replace `older` with `mobile`. 

Alternatively, I might express this as 

> The GPU wants to run like a firehose, and a render target switch forces the GPU to redirect that stream temporarily and then direct it back again. On mobile GPUs this is particularly disruptive to rendering throughput.

Reply:
Love it!